agent/lib/kontena/observable.rb

```diff
@@ -43,33 +68,48 @@ def reset_observable
       notify_observers
     end
 
-    # Observer actor is observing this Actor's @value.
-    # Updates to value will send to update_observe on given actor.
-    # Returns current value.
+    # Observer actor is observing this Actor's @observable_value.
+    # Subscribes actor for updates to our observable value, sending Kontena::Observable::Message to given actor.mailbox.
+    # Links actor, such that observable actor crashes propagate to the observer.
+    # Returns Kontena::Observable::Message with current value.
+    # The observing actor will be unsubscribed/unlinked once the actor.mailbox or observe becomes !alive?.
     #
-    # @param observer [Celluloid::Proxy::Cell<Observer>]
+    # @param actor [Celluloid::Proxy::Cell<Actor>]
     # @param observe [Observer::Observe]
-    # @return [Object, nil] possible existing value
-    def add_observer(observer, observe)
-      debug "observer: #{observe} <- #{@observable_value.inspect[0..64] + '...'}"
+    # @return [Kontena::Observable::Message] with current value
+    def add_observer(actor, observe)
+      if Celluloid.current_actor != actor
+        links = Celluloid.links
+      end
+
+      if !observable?
+        debug "observer: #{observe.describe_observer}..."
 
-      observers[observe] = observer
+        links << actor if links
+        observers[observe] = actor
+
+      else
+        debug "observer: #{observe.describe_observer} <= #{@observable_value.inspect[0..64]}..."
+
+        links << actor if links
+        observers[observe] = actor
+      end
 
-      return @observable_value
+      return Message.new(observe, self, @observable_value)
     end
 
-    # Update @value to each Observer::Observe
+    # Update @observable_value to each Kontena::Observer::Observe
     def notify_observers
-      observers.each do |observe, observer|
-        if observer.alive?
-          debug "notify: #{observe} <- #{@observable_value.inspect[0..64] + '...'}"
+      observers.each do |observe, actor|
+        if observe.alive? && actor.mailbox.alive?
+          debug "notify: #{observe.describe_observer} <- #{@observable_value.inspect[0..64]}"
 
-          # XXX: is the Observable's Celluloid.current_actor guranteed to match the Actor[:node_info_worker] Celluloid::Proxy::Cell by identity?
-          observer.async.update_observe(observe, Celluloid.current_actor, @observable_value)
+          actor.mailbox << Message.new(observe, self, @observable_value)
         else
-          debug "observer died: #{observe}"
+          debug "dead: #{observe.describe_observer}"
 
           observers.delete(observe)
+          Celluloid.links.delete(actor)
```

Comment:
Might need something to actually check the `observe.alive?` when the Observable actor shuts down, but extending the actor shutdown/cleanup logic from a mixin like `Kontena::Observable` is questionable... maybe with some hack to store the `Observe` instead of the calling actor in the links.